### ISSUE TITLE
Maybe should be parametric

### ISSUE THREAD

**Issue by jack-williams** [gigobyte/purify#62]
The maybe implementation does not satisfy the monad laws because the implementation is not parametric in the wrapped value. Example tests that should pass:
```ts
test('parametricity', () => {
  expect(Just(undefined).isNothing()).toEqual(false)
  expect(Just(null).isNothing()).toEqual(false)
})

test('Left Id', () => {
  expect(Just(undefined).chain(x => Just(typeof x === 'undefined'))).toEqual((x => Just(typeof x === 'undefined'))(undefined))
  expect(Just(null).chain(x => Just(typeof x === 'object'))).toEqual((x => Just(typeof x === 'object'))(null))
})
```

**Comment by gigobyte** [comment]
That's really funny. I am, at the moment, working on fixing exactly this problem.
This is more of a implementation defect, as `Just(null)` is the same as `Nothing` [internally](https://github.com/gigobyte/purify/blob/master/src/adts/Maybe.ts#L122).

**Comment by jack-williams** [comment]
Closed by 6249fbf7085a32dc2cc3ecfc9e6d74e975c15d0c
